include/alpaka/intrinsic/Traits.hpp

```diff
@@ -0,0 +1,81 @@
+/* Copyright 2020 Sergei Bastrakov
+ *
+ * This file is part of Alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#include <alpaka/core/Common.hpp>
+#include <alpaka/core/Concepts.hpp>
+
+#include <type_traits>
+
+namespace alpaka
+{
+    //-----------------------------------------------------------------------------
+    //! The intrinsic specifics
+    namespace intrinsic
+    {
+        struct ConceptIntrinsic{};
+
+        //-----------------------------------------------------------------------------
+        //! The intrinsics traits.
+        namespace traits
+        {
+            //#############################################################################
+            //! The popcount trait.
+            template<
+                typename TWarp,
+                typename TSfinae = void>
+            struct Popcount;
```

Comment:
As we did for all other traits e.g. math traits we should add the argument type to the `class`. This will than remove also the requirement of two type overloaded free entry function `popcount`.
Having the type in the trait class will give us also the possibility to specialize the function for 16bit or 8bit, if in the future available, without a need of a cast.

Reply:
I thought about it. But I decided not to do it, as now popcount explicitly works for two types only. This is not like `cos` for example that one could call for any type, including user-defined ones.

Reply:
So to reiterate, I could do a template parameter there. But i think this is misleading, as the type is not supposed to be generic, it is strictly one of the two supported types. And so I think it's better to be explicit. In terms of code line number both are about the same, as it's two template specialization with one function vs. non-template with two function overloads.

Reply:
but there is also a `__popcnt16` available. Maybe only for the MS compiler.

Reply:
Another point for having the type in the template is that we are working on getting vector types into alpaka, these will benefit from have it in the template signature. It can also be refactored later.

@BenjaminW3 Do you have a preference here?

Reply:
The common subset is 32-bit and 64-bit only, all of CUDA, HIP and gcc only support these two versions.

Reply:
But generally I'm not against your proposal, think it's also fine, but just slightly worse than my version.

Reply:
let us wait for a third opinion. I am not fixed to my suggestion.

Reply:
In general I would prefer the template parameter version but in this special case I am also fine with the current solution. We can still change it later when required.